Automatic Hit Selection

labkey-alan opened the pull request:
#### Rationale
This PR adds the client side code necessary to configure automatic hit selection for plate based assays.

#### Related Pull Requests
- https://github.com/LabKey/labkey-api-js/pull/189
- https://github.com/LabKey/labkey-ui-components/pull/1649
- https://github.com/LabKey/labkey-ui-premium/pull/604
- https://github.com/LabKey/platform/pull/6123
- https://github.com/LabKey/limsModules/pull/941

#### Changes
- Add HitCriteriaRenderer
- Add HitCriteriaModal
- Add useLoadableState
  - Helper hook that takes a loader method and returns loadingState, value, and error
- AssayProtocolModel: Add hitCriteria
- Render Hit Criteria components in AssayDesignerPanels

labkey-alan committed:
Fix unique key warnings in domain designer

labkey-alan committed:
CLean up filter panel styles

labkey-alan committed:
Add HitCriteriaRenderer

labkey-alan committed:
Add HitCriteriaModal and useLoadableState

labkey-alan committed:
Render HitCriteria

- Render HitCriteriaModal
- Add AssayDomainForm, which allows us to memoize props

labkey-alan committed:
Update release notes

labkey-alan committed:
FilterExpressionView: memoize callbacks

labkey-alan committed:
AssayDesignerPanels: Fix issue with appDomainHeaders, fix tests

labkey-alan committed:
Cleanup

labkey-nicka committed:
Prevent serialization of "hitCriteria" (stop gap)

labkey-alan committed:
Bump @labkey/api

labkey-alan committed:
Add request util

labkey-alan committed:
Add getFilterCriteriaColumns to assay API wrapper

labkey-alan committed:
Update AssayDesigner to assume filterCriteria is on DomainFields

labkey-alan committed:
Remove protocolModel from FilterCriteriaContext

labkey-alan committed:
Remove hack

labkey-alan committed:
FilterCriteriaModal - use propertyIds

labkey-alan committed:
request: Add RequesstHandler

labkey-alan committed:
Fix styling for FilterCriteriaRenderer

labkey-alan committed:
Update filter criteria when field name changes

labkey-alan committed:
FilterCriteriaModal: make FilterCriteriaMap a JS map

labkey-alan committed:
Export FilterCriteriaRenderer

labkey-alan committed:
FilterCriteriaRenderer: optionally render empty message

labkey-alan committed:
Fix styling for hit-criteria-renderer

labkey-alan committed:
Fix styling for FilterCriteriaRenderer

labkey-alan committed:
Fix class names

labkey-alan committed:
FilterCriteriaModal: Add empty messages

Fix issue when there are no fields

labkey-alan committed:
AssayDesignerPanels: Fix issue with saveFilterCriteria

labkey-alan committed:
FilterCriteriaModal: Give reference properties stable propertyId values

labkey-alan committed:
FilterCriteriaModal: filter fields to openTo and related

labkey-alan committed:
DomainField.serialize - Don't change propertyId or referencePropertyId in filterCriteria

labkey-alan committed:
AssayPropertiesInput: Disable Edit Criteria button when plate metadata is not enabled

labkey-alan committed:
AssayDesignerPanels: Hide field "edit criteria" buttons when plate metadata is disabled

labkey-alan committed:
FilterCriteriaModal.tsx: Fix issue with empty fields

labkey-alan committed:
Bump api-js

labkey-alan committed:
request.ts - Fix types

labkey-alan committed:
models.test.ts - Fix tests

labkey-nicka commented on packages/components/src/internal/FilterCriteriaModal.tsx:
```diff
@@ -0,0 +1,201 @@
+                    isKeyField: false,
+                }))
+            );
+        }, [] as FilterCriteriaField[]);
```
nit: Instead of `as FilterCriteriaField[]` consider `.reduce<FilterCriteriaField[]>`.

labkey-nicka commented on packages/components/src/internal/FilterCriteriaModal.tsx:
```diff
@@ -0,0 +1,201 @@
+            })
+            .map(field => field.name)
+            .toArray();
+        const referenceFields = await fetch(protocolId, sourceFields, container);
```
Currently, if source fields is empty this still makes the request to the endpoint which in turn responds with a 400 because at least one column is required. Perhaps we can short-circuit here and just return an empty array if `sourceFields` turns up empty.

labkey-nicka commented on packages/components/src/internal/FilterCriteriaModal.tsx:
```diff
@@ -0,0 +1,201 @@
+export const FilterCriteriaModal: FC<Props> = memo(({ onClose, onSave, openTo, protocolModel }) => {
+    const { api } = useAppContext();
+    const { protocolId, container } = protocolModel;
+    const domain = useMemo(
```
nit: Consider using `getDomainByNameSuffix()`.

```tsx
const domain = useMemo(() => protocolModel.getDomainByNameSuffix('Data'), [protocolModel]);
```

labkey-nicka commented on packages/components/src/internal/FilterCriteriaRenderer.tsx:
```diff
@@ -0,0 +1,54 @@
+    field: DomainField;
+}
+
+const FilterCriteriaField: FC<FieldWithCriteria> = memo(({ field }) => {
```
Adding `displayName` for components helps in stacktraces to identify which functional component is causing a problem. Consider adding `displayName` for all components declared in this PR.

labkey-nicka commented on packages/components/src/internal/FilterCriteriaRenderer.tsx:
```diff
@@ -0,0 +1,54 @@
+
+import { DomainField } from './components/domainproperties/models';
+
+function getFilterDisplaySymbol(op: string) {
```
I recommend utilizing `Filter.getFilterTypeForURLSuffix(op).getDisplaySymbol()`

labkey-nicka commented on packages/components/src/internal/components/domainproperties/DomainRowExpandedOptions.tsx:
```diff
@@ -269,6 +270,9 @@ export class DomainRowExpandedOptions extends React.Component<Props> {
             domainFormDisplayOptions,
             getDomainFields,
         } = this.props;
+        const dataType = field.dataType.name;
```
Create a consolidated function house the logic of what is a valid field for filter criteria. Then utilize this function in the `FilterCriteriaModal.fieldLoaderFactory` as well.

labkey-nicka commented on packages/components/src/internal/components/domainproperties/FieldFilterCriteria.tsx:
```diff
@@ -0,0 +1,43 @@
+    const context = useFilterCriteriaContext();
+    const openModal = context?.openModal;
+    const onClick = useCallback(() => openModal(propertyId), [openModal, propertyId]);
+    const fields = useMemo(() => [field], [field]);
```
Gotta love React.

labkey-alan committed:
Address PR feedback

labkey-alan committed:
DomainForm: Move toolbar code to separate component

labkey-alan committed:
Add filterCriteriaToStr

labkey-alan committed:
DomainPropertiesGrid: optionally render Filter Criteria

labkey-alan committed:
AssayDesignerPanels.tsx: fix issue with header prefix

labkey-alan committed:
PlateMetadataInput - Only render filter criteria for valid fields

labkey-alan committed:
FilterCriteriaModal: fix issues

- Fix issue with openTo looking at wrong fields
- Fix issue with filter types not being properly limited to the field type

labkey-alan committed:
Bump api-js

labkey-alan committed:
Prep for release
